### PR TITLE
OOP Principles / it-academy / JC1-117-22 / Victor Polonnikov / Done Tasks 1-2

### PR DESCRIPTION
**Pull request by VictPol**
## Pull request state

Actual status for PR:
- [x] implemented
- [x] reviewed
- [x] fixed
- [ ] approved

### CODE REVIEW

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Warrior.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Warrior extends Hero {
+
+    Warrior (String name) {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/Warrior.java:5]:
```suggestion
public class Warrior extends Hero {
    public Warrior (String name) {
```

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    @Test
+    public void testArcher(){
+        Assertions.assertEquals("Bob", archer.getName());
+    }
+
+    @Test
+    public void testArcherAttackEnemy() {
+        archer.attackEnemy();
+        System.out.println("Archer attacks enemy");
+    }
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java:19]: it isn't test

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        Assertions.assertEquals(false, archer.isAlive());
+    }
+
+    @Test
+    public void testEnemyAttackArcher() {
+        enemy.attackHeroes(archer);
+        Assertions.assertEquals(190, archer.getHealth());
+    }
+
+}
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java:51]:
```suggestion
    }
}
```

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    @Test
+    public void testCat() {
+        Assertions.assertEquals("Tosha", cat.getName());
+    }
+
+    @Test
+    public void testMew() {
+        cat.mew();
+        System.out.println("mjau");
+    }
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java:19]: it isn't

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/HeroTests.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        hero.attackEnemy();
+        System.out.println("Hero attacks enemy");
+    }
+
+    @Test
+    public void testHeroAttackEnemyAndDamage() {
+        hero.attackEnemy(enemy);
+        Assertions.assertEquals(95, enemy.getHealth());
+    }
+}*/
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/HeroTests.java:27]: remove unused code

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/MagTests.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class MagTests {
+    Mag mag = new Mag("Jonh");
+    Enemy enemy = new Enemy(200);
+
+    @Test
+    public void testMag(){
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/MagTests.java:11]:
```suggestion
    public void testMag() {
```

**Review by rakovets — CHANGES_REQUESTED** [review]
You need to fix and add correct solution

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Archer.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Archer extends  Hero {
+    Archer(String name) {
+        super(name);
+    }
+
+    @Override
+    public void attackEnemy() {
+        System.out.println("Archer attacks enemy");
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Archer.java:10]: This method can`t be rmoved, it is declared as abstract method in syperclass Hero according tj the specification and must be overrided in subclass Archer

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Cat.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Cat {
+    private String name;
+
+    Cat(String name) {
+        this.name = name;
+    }
+
+    public void mew() { System.out.println("mjau"); }
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Cat.java:10]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Cat.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+    public void mew(Person person) {
+        person.changeHappiness(-1);
+        this.mew();
+    }
+
+    public void purr() {
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Cat.java:17]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Cat.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    private String name;
+
+    Cat(String name) {
+        this.name = name;
+    }
+
+    public void mew() { System.out.println("mjau"); }
+
+    public void mew(Person person) {
+        person.changeHappiness(-1);
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Cat.java:13]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Enemy.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Enemy implements Mortal {
+    private int health;
+
+    Enemy(int health) {
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Enemy.java:6]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public abstract class Hero implements Mortal {
+    private String name;
+    private int health;
+
+    {
+        health = 200;
+    }
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java:9]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    public String getName() {
+        return this.name;
+    }
+
+    public abstract void attackEnemy();
+
+    public void attackEnemy(Enemy enemy) {
+        enemy.takeDamage(5);
+    }
+    public void setHealth(int health) {
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java:24]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        this.health = health;
+    }
+
+    public int getHealth(){
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java:28]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        return health;
+    }
+    public void takeDamage(int damage) {
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Hero.java:31]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Mag.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Mag extends Hero {
+    Mag(String name) {
+        super(name);
+    }
+
+    @Override
+    public void attackEnemy() {
+        System.out.println("Mag attacks enemy");
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Mag.java:10]: This method can`t be rmoved, it is declared as abstract method in syperclass Hero according to the specification and must be overrided in subclass Mag

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Persian.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Persian extends Cat {
+
+    Persian(String name) {
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Persian.java:5]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Persian.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Override
+    public void mew(Person person) {
+        person.changeHappiness(-2);
+        this.mew();
+    }
+
+    @Override
+    public  void purr() {
+        System.out.println("murrPersian");
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Persian.java:22]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Person.java [hunk 19]
```diff
@@ -0,0 +1,20 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Person {
+
+    final int HUNDRED_PERCENTS = 100;
+
+    private double happiness;
+
+    Person(double happiness) {
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Person.java:9]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Person.java [hunk 20]
```diff
@@ -0,0 +1,20 @@
+package com.rakovets.course.java.core.practice.oop_principles;
+
+public class Person {
+
+    final int HUNDRED_PERCENTS = 100;
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Person.java:5]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Warrior.java (hunk 1 above)
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Warrior.java:5]: fixed

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java (hunk 3 above)
Comment (VictPol) [reply src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java:51]: fixed

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/HeroTests.java (hunk 5 above)
Comment (VictPol) [reply src/test/java/com/rakovets/course/java/core/practice/oop_principles/HeroTests.java:27]: fixed

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/MagTests.java (hunk 6 above)
Comment (VictPol) [reply src/test/java/com/rakovets/course/java/core/practice/oop_principles/MagTests.java:11]: fixed

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/Persian.java [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+public class Persian extends Cat {
+
+    Persian(String name) {
+        super(name);
+    }
+
+    @Override
+    public void mew() {
+        System.out.println("mjauPersian");
```
[the remark this answers is not in the document]
Comment (VictPol) [reply src/main/java/com/rakovets/course/java/core/practice/oop_principles/Persian.java:11]: fixed

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java (hunk 2 above)
Comment (VictPol) [reply src/test/java/com/rakovets/course/java/core/practice/oop_principles/ArcherTests.java:19]: fixed

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java (hunk 4 above)
Comment (VictPol) [reply src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java:19]: fixed

**Comment by rakovets** [comment]
outdated

**Pull request closed without merging**